File: solutions/java/bird-watcher/1/src/main/java/BirdWatcher.java
class BirdWatcher {
    private final int[] birdsPerDay;

    public BirdWatcher(int[] birdsPerDay) {
        this.birdsPerDay = birdsPerDay.clone();
    }

    // Task 1: Return last week's fixed counts
    public int[] getLastWeek() {
        return new int[] { 0, 2, 5, 3, 7, 8, 4 };
    }

    // Task 2: Return today's count (last element of array)
    public int getToday() {
        if (birdsPerDay.length == 0) return 0;
        return birdsPerDay[birdsPerDay.length - 1];
    }

    // Task 3: Increment today's count
    public void incrementTodaysCount() {
        if (birdsPerDay.length > 0) {
            birdsPerDay[birdsPerDay.length - 1]++;
        }
    }

    // Task 4: Check if any day had zero birds
    public boolean hasDayWithoutBirds() {
        for (int count : birdsPerDay) {
            if (count == 0) {
                return true; // Early exit: found a zero
            }
        }
        return false;
    }

    // Task 5: Sum birds for first N days (with bounds safety)
    public int getCountForFirstDays(int numberOfDays) {
        int total = 0;
        int limit = Math.min(numberOfDays, birdsPerDay.length);
        for (int i = 0; i < limit; i++) {
            total += birdsPerDay[i];
        }
        return total;
    }

    // Task 6: Count days with 5+ birds (busy days)
    public int getBusyDays() {
        int busyDays = 0;
        for (int count : birdsPerDay) {
            if (count >= 5) {
                busyDays++;
            }
        }
        return busyDays;
    }
}

File: solutions/r/bird-watcher/1/bird-watcher.R
today <- function(birds_per_day) {
  # Return the last element (today's count)
  birds_per_day[length(birds_per_day)]
}

increment_todays_count <- function(birds_per_day) {
  # Increment the last element by 1
  birds_per_day[length(birds_per_day)] <- birds_per_day[length(birds_per_day)] + 1
  birds_per_day
}

has_day_without_birds <- function(birds_per_day) {
  # Check if any day had 0 birds
  any(birds_per_day == 0)
}

count_for_first_days <- function(birds_per_day, num_days) {
  # Sum the first num_days elements
  sum(birds_per_day[1:num_days])
}

busy_days <- function(birds_per_day) {
  # Count days with 5 or more birds
  sum(birds_per_day >= 5)
}

running_total <- function(birds_per_day) {
  # Calculate cumulative sum
  cumsum(birds_per_day)
}

busy_days_of_week <- function(birds_per_day, day_names) {
  # Sort day names by bird count in descending order
  day_names[order(birds_per_day, decreasing = TRUE)]
}